### PR TITLE
Use a ScaleTransform to prevent text clipping in the outlining tag hint (fixes #2128)

### LINKED ISSUE #2128: Outlining cuts off closing }
Ported from TFS WorkItem: **1037510**

---
## outlining

The closing } in the preview is cutoff half through.

### PR DESCRIPTION
**Pull request by tuespetre**
Fixes #2128 

💻 🖱 ⌨️

### CODE REVIEW

**Comment by tuespetre** [comment]
@Pilchie Here we are! 😋

File: src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
 using System;
+using System.Windows.Media;
 using Microsoft.VisualStudio.Text.Editor;
 
 namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
 {
     internal static class IWpfTextViewExtensions
     {
-        public static void SizeToFit(this IWpfTextView view)
+        public static void SizeToContentsAndScale(this IWpfTextView view, double scaleFactor = 0.75)
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs]: Nit: We don't normally make parameters optional in things that aren't part of the public API.

File: src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
         {
             // Computing the height of something is easy.
             view.VisualElement.Height = view.LineHeight * view.TextBuffer.CurrentSnapshot.LineCount;
+            
+            // view.ZoomLevel can cause clipping so use a ScaleTransform instead.
+            view.VisualElement.LayoutTransform = new ScaleTransform(scaleFactor, scaleFactor);
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs]: @olegtk @dpugh - any ideas on *why* `view.ZoomLevel` might cause clipping?

**Comment by Pilchie** [comment]
Looks okay to me, but would like @jasonmalinowski  and/or @dpoeschl to chime in too for any thoughts on HDPI/etc.

File: src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs (hunk 2 above)
Comment (sharwell) [inline src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs]: :arrow_right: The `IWpfTextView.ZoomLevel` property eventually gets used to calculate the height of lines through the old [`IVsTextView.GetLineHeight`](https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.textmanager.interop.ivstextview.getlineheight.aspx) API, which returns an integer height for each line. Users who have a font size in the editor that is not divisible by 4 will see pixels trimmed from the bottom of this view. Set your font size to 12 for example and the bug does not appear.

Comment (sharwell) [inline src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs]: 💭 Actually in this case the problem may lie in the assumptions made for the use of `view.LineHeight`.

File: src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
 using Microsoft.VisualStudio.Text.Editor;
 
 namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
 {
     internal static class IWpfTextViewExtensions
     {
-        public static void SizeToFit(this IWpfTextView view)
+        public static void SizeToContentsAndScale(this IWpfTextView view, double scaleFactor = 0.75)
         {
             // Computing the height of something is easy.
```
Comment (sharwell) [inline src/EditorFeatures/Core/Shared/Extensions/IWpfTextViewExtensions.cs]: 💭 Whoever wrote this was wrong 😆

**Comment by CyrusNajmabadi** [comment]
> Users who have a font size in the editor that is not divisible by 4 will see pixels trimmed from the bottom of this view. Set your font size to 12 for example and the bug does not appear.

Is this fixable by just rounding up then?

**Comment by CyrusNajmabadi** [comment]
This has been reported by another customer.  Can we shepherd this change in?

**Comment by sharwell** [comment]
Superseded by #26551.

**Pull request closed without merging**